File: Core/resource_monitor/broker/service.py
"""Service entrypoint: register HTTP routes + pipe actions, start the pipe alias + workers.

Manifest writing lives in ``Core/resource_monitor/run.py`` per the
service-owns-manifest convention; ``install()`` only wires the HTTP
routes, pipe action handlers, and pipe surface that the broker exposes.

The Flask routes and pipe actions both dispatch into the same
underlying helpers (``_try_grant``, ``_registry``, ``_state_snapshot``,
``_model_cache``), so there is exactly one implementation of each
operation regardless of which transport the caller used. The action
surface (`vram.reserve`, `vram.release`, ...) matches the Rust port at
``rust/crates/wylde-vram-broker/src/service.rs`` byte-for-byte, so the
``WYLDE_WYLDE_VRAM_BROKER_IMPL`` strangler-fig switch is safe to flip
without breaking Python clients.
"""

from __future__ import annotations

import logging
import time
from typing import Any, Dict, List

from flask import Flask, Response, jsonify, request

from Core.shared.ipc import IpcError, register_action, unregister_action

from .config import _DEFAULT_TTL, _MODEL_CACHE_TTL_S
from .model_cache import _model_cache
from .policy import _try_grant
from .registry import _init_nvml, _refresh_nvml, _registry
from .workers import _start_background, _state_snapshot

logger = logging.getLogger(__name__)

# Action names exposed to pipe callers. Kept as a module-level tuple so
# ``_reset_for_tests`` can sweep the registry without re-listing each name.
_ACTION_NAMES = (
    "vram.reserve",
    "vram.release",
    "vram.heartbeat",
    "vram.state",
    "vram.leases",
    "vram.cache",
    "vram.evict",
)


def _start_pipe_alias(app: Flask) -> None:
    """Start a PipeServer for 'vram-broker' backed by the Flask app.

    Creates ``\\\\.\\pipe\\wylde-vram-broker`` so the Tauri list_pipes
    call returns 'vram-broker' and the dashboard shows the broker as
    active. All /vram/* requests routed through this pipe reach the
    same handlers the Flask app serves on its HTTP port.
    """
    try:
        from Core.shared import ipc

        server = ipc.PipeServer("vram-broker", app)
        server.start()
        logger.info("vram_broker: pipe alias started (\\\\.\\pipe\\wylde-vram-broker)")
    except Exception as e:
        logger.warning("vram_broker: pipe alias start failed: %s", e)


def install(app: Flask, gpu_available: bool = True) -> None:
    """Register /vram/* routes on `app` and start background threads.

    Safe to call more than once; subsequent calls are no-ops so the
    boot wrapper can call install() during initial boot and again
    after a late nvml probe without double-registering routes.
    """
    if getattr(app, "_vram_broker_installed", False):
        return
    app._vram_broker_installed = True  # type: ignore[attr-defined]

    if gpu_available:
        _init_nvml()
        _refresh_nvml()

    @app.route("/vram/reserve", methods=["POST"])
    def _vram_reserve() -> tuple[Response, int]:
        req = request.get_json(silent=True) or {}
        result = _try_grant(req)
        if result.get("ok"):
            return jsonify(result["lease"]), 200
        err = result["error"]
        # Map domain codes to HTTP — 409 for "can't fit now", 400 for bad input.
        status = 400 if err.get("code") == "invalid_request" else 409
        return jsonify(
            {
                "error": err.get("message"),
                "code": err.get("code"),
                "details": err.get("details", {}),
            }
        ), status

    @app.route("/vram/release", methods=["POST"])
    def _vram_release() -> tuple[Response, int]:
        body = request.get_json(silent=True) or {}
        lease_id = str(body.get("lease_id", ""))
        lease = _registry.remove(lease_id)
        if lease is None:
            return jsonify({"ok": True, "known": False}), 200
        logger.info(
            "vram_broker: release lease=%s service=%s model=%s",
            lease.lease_id[:8],
            lease.service,
            lease.model,
        )
        return jsonify({"ok": True, "known": True, "freed_bytes": lease.bytes}), 200

    @app.route("/vram/heartbeat", methods=["POST"])
    def _vram_heartbeat() -> tuple[Response, int]:
        body = request.get_json(silent=True) or {}
        lease_id = str(body.get("lease_id", ""))
        ttl = float(body.get("ttl", _DEFAULT_TTL))
        lease = _registry.touch(lease_id, ttl)
        if lease is None:
            return jsonify(
                {"error": "lease not found or already reaped", "code": "not_found"}
            ), 404
        return jsonify(
            {"lease_id": lease.lease_id, "expires_at": lease.expires_at}
        ), 200

    @app.route("/vram/state", methods=["GET"])
    def _vram_state() -> tuple[Response, int]:
        _refresh_nvml()
        return jsonify(_state_snapshot()), 200

    @app.route("/vram/leases", methods=["GET"])
    def _vram_leases() -> tuple[Response, int]:
        return jsonify(
            {"leases": [lease.to_wire() for lease in _registry.all_leases()]}
        ), 200

    @app.route("/vram/cache", methods=["GET"])
    def _vram_cache() -> tuple[Response, int]:
        """Return the recently-used (service, model) cache.

        Used by services to decide whether to skip cold-start work, if
        their model is in the warm cache, they should reuse it. The cache
        is purely informational; cold/warm decisions still go through
        /vram/reserve.
        """
        entries = _model_cache.all()
        return jsonify(
            {
                "ttl_s": _MODEL_CACHE_TTL_S,
                "entries": [
                    {
                        "service": e.service,
                        "model": e.model,
                        "bytes": e.bytes,
                        "last_used": e.last_used,
                        "warm_for": max(
                            0.0, _MODEL_CACHE_TTL_S - (time.time() - e.last_used)
                        ),
                    }
                    for e in sorted(entries, key=lambda x: -x.last_used)
                ],
            }
        ), 200

    @app.route("/vram/evict", methods=["POST"])
    def _vram_evict_broker() -> tuple[Response, int]:
        """Manual admin eviction: drop a lease by id as if it had timed out.
        Used by the GUI's 'unload' button and by tests. Does NOT signal the
        owning service — this is a force-drop, so the service must have
        already released its model externally (or the user accepts that
        accounting will drift until the service exits)."""
        body = request.get_json(silent=True) or {}
        lease_id = str(body.get("lease_id", ""))
        lease = _registry.remove(lease_id)
        if lease is None:
            return jsonify({"ok": False, "error": "not_found"}), 404
        return jsonify({"ok": True, "freed_bytes": lease.bytes}), 200

    _register_actions()
    _start_pipe_alias(app)
    _start_background()


def _register_actions() -> None:
    """Bind every ``vram.*`` action onto the shared IPC registry.

    Handlers call the same underlying helpers the Flask routes use, so
    the route and action surfaces stay byte-equivalent. Errors are
    surfaced via :class:`IpcError` so the dispatcher emits a structured
    ``{ok:False, error:{code, message, details}}`` reply that matches the
    Rust impl's ``Reply::err`` shape.
    """
    register_action("vram.reserve", _action_reserve)
    register_action("vram.release", _action_release)
    register_action("vram.heartbeat", _action_heartbeat)
    register_action("vram.state", _action_state)
    register_action("vram.leases", _action_leases)
    register_action("vram.cache", _action_cache)
    register_action("vram.evict", _action_evict)


def _action_reserve(payload: Any) -> Dict[str, Any]:
    """Request a VRAM lease; may preempt lower-priority holders."""
    req = payload if isinstance(payload, dict) else {}
    result = _try_grant(req)
    if result.get("ok"):
        # Return the bare lease dict, with optional preempted /
        # soft_eviction / dedup as sibling fields. Matches the Rust
        # handler's shape so cross-impl callers see identical payloads.
        out: Dict[str, Any] = dict(result["lease"])
        for key in ("preempted", "soft_eviction", "dedup"):
            if key in result:
                out[key] = result[key]
        return out
    err = result["error"]
    raise IpcError(
        str(err.get("code", "unknown")),
        str(err.get("message", "reserve failed")),
        err.get("details") or None,
    )


def _action_release(payload: Any) -> Dict[str, Any]:
    """Release a lease by id; idempotent."""
    body = payload if isinstance(payload, dict) else {}
    lease_id = str(body.get("lease_id", ""))
    lease = _registry.remove(lease_id)
    if lease is None:
        return {"ok": True, "known": False}
    logger.info(
        "vram_broker: release lease=%s service=%s model=%s",
        lease.lease_id[:8],
        lease.service,
        lease.model,
    )
    return {"ok": True, "known": True, "freed_bytes": lease.bytes}


def _action_heartbeat(payload: Any) -> Dict[str, Any]:
    """Extend a lease's expiry by ttl seconds."""
    body = payload if isinstance(payload, dict) else {}
    lease_id = str(body.get("lease_id", ""))
    ttl = float(body.get("ttl", _DEFAULT_TTL))
    lease = _registry.touch(lease_id, ttl)
    if lease is None:
        raise IpcError("not_found", "lease not found or already reaped")
    return {"lease_id": lease.lease_id, "expires_at": lease.expires_at}


def _action_state(_payload: Any) -> Dict[str, Any]:
    """Snapshot of GPU totals, leases, model cache, and config."""
    _refresh_nvml()
    return _state_snapshot()


def _action_leases(_payload: Any) -> Dict[str, Any]:
    """List all live leases."""
    return {"leases": [lease.to_wire() for lease in _registry.all_leases()]}


def _action_cache(_payload: Any) -> Dict[str, Any]:
    """List the soft-LRU (service, model) keep-warm cache."""
    now = time.time()
    entries: List[Dict[str, Any]] = [
        {
            "service": e.service,
            "model": e.model,
            "bytes": e.bytes,
            "last_used": e.last_used,
            "warm_for": max(0.0, _MODEL_CACHE_TTL_S - (now - e.last_used)),
        }
        for e in sorted(_model_cache.all(), key=lambda x: -x.last_used)
    ]
    return {"ttl_s": _MODEL_CACHE_TTL_S, "entries": entries}


def _action_evict(payload: Any) -> Dict[str, Any]:
    """Force-drop a lease by id; does NOT signal the owning service."""
    body = payload if isinstance(payload, dict) else {}
    lease_id = str(body.get("lease_id", ""))
    lease = _registry.remove(lease_id)
    if lease is None:
        raise IpcError("not_found", "lease not found")
    return {"ok": True, "freed_bytes": lease.bytes}


def stop() -> None:
    from .workers import _threads

    _threads.stop.set()
    try:
        from Core.shared.manifest import stop_heartbeat

        stop_heartbeat("vram-broker")
    except Exception:
        pass


def _reset_for_tests() -> None:
    """Exposed for tests so they can reset state between cases. Delegates to
    the per-submodule resets — keeps the registry / threads bindings stable
    so the shim's re-exports stay live."""
    from . import registry, workers

    stop()
    for name in _ACTION_NAMES:
        unregister_action(name)
    registry._reset()
    workers._reset()

File: rust/crates/wylde-vram-broker/src/service.rs
//! Service entrypoint: register pipe action handlers and start workers.
//!
//! Rust port of `Core/resource_monitor/broker/service.py`. The Python
//! version wired Flask HTTP routes and bridged them onto a pipe via
//! `PipeServer(app)`; the Rust version skips the HTTP surface and exposes
//! the same operations directly as pipe-side actions on the shared
//! `wylde_shared::ipc` server.
//!
//! Action surface (one-to-one with the Python `/vram/*` routes):
//!
//! | action            | python route       | data shape                                       |
//! |-------------------|--------------------|--------------------------------------------------|
//! | `vram.reserve`    | `POST /vram/reserve`   | `{service,model,bytes,priority,ttl,preempt,pid,client_nonce}` |
//! | `vram.release`    | `POST /vram/release`   | `{lease_id}`                                     |
//! | `vram.heartbeat`  | `POST /vram/heartbeat` | `{lease_id, ttl}`                                |
//! | `vram.state`      | `GET  /vram/state`     | `{}`                                             |
//! | `vram.leases`     | `GET  /vram/leases`    | `{}`                                             |
//! | `vram.cache`      | `GET  /vram/cache`     | `{}`                                             |
//! | `vram.evict`      | `POST /vram/evict`     | `{lease_id}`                                     |

use std::sync::atomic::{AtomicBool, Ordering};

use serde_json::{json, Value};
use wylde_shared::ipc::{register_action_with_meta, unregister_action, IpcError, Reply};

use crate::config::Config;
use crate::inventory;
use crate::model_cache::model_cache;
use crate::policy::{self, GrantRequest, GrantResult};
use crate::registry::{init_nvml, refresh_nvml, refresh_sysinfo, registry};
use crate::time::now_secs;
use crate::workers;

static INSTALLED: AtomicBool = AtomicBool::new(false);

/// Register every `vram.*` action on the process-wide pipe registry and
/// start the background workers. Idempotent — repeat calls are no-ops so
/// boot wrappers can call `install` once at startup and again after a
/// late NVML probe without double-registering.
pub fn install(gpu_available: bool) {
    if INSTALLED.swap(true, Ordering::SeqCst) {
        return;
    }
    if gpu_available {
        init_nvml();
        refresh_nvml();
        // Populate DRAM totals at boot too, not just on the reaper's first
        // tick: a model larger than VRAM needs the spillover capacity to be
        // known when the very first `vram.reserve` lands, or it spuriously
        // hits `would_exceed_total`. The reaper keeps refreshing on its
        // cadence. Gated with the GPU probe so `install(false)` keeps tests
        // hermetic (they control DRAM via `registry().set_dram`).
        refresh_sysinfo();
    }

    register_action_with_meta(
        "vram.reserve",
        |payload: Value| async move { handle_reserve(payload).await },
        // Reply shape (lease): lease_id, service, model, bytes (VRAM portion),
        // dram_bytes (DRAM portion; 0 unless spilled), priority, granted_at,
        // expires_at, heartbeat_at, pid, synthetic, estimated, client_nonce.
        // Sibling fields when relevant: preempted, soft_eviction, dedup,
        // spilled. When `bytes` is omitted or <= 0 the broker estimates from
        // the keep-warm cache, observed synthetic leases, or name heuristics.
        "Request a VRAM lease; may spill into DRAM and/or preempt lower-priority holders.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.release",
        |payload: Value| async move { handle_release(payload).await },
        "Release a lease by id; idempotent. Reply: {ok, known, freed_bytes}.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.heartbeat",
        |payload: Value| async move { handle_heartbeat(payload).await },
        "Extend a lease's expiry by ttl seconds. Reply: {lease_id, expires_at}.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.state",
        |payload: Value| async move { handle_state(payload).await },
        // Snapshot shape:
        //   gpu: { total_bytes, actual_used_bytes, reserved_bytes,
        //          free_for_grant, safety_margin, name, nvml_fresh_s }
        //   system_memory: { total_bytes, actual_used_bytes,
        //          reserved_dram_bytes, free_for_grant_dram,
        //          safety_margin, sysinfo_fresh_s }
        //   leases: [Lease { ..., bytes, dram_bytes, estimated, ... }]
        //   by_service: [{ service, bytes, dram_bytes, count, priority,
        //          synthetic }]
        //   model_cache: { ttl_s, entries: [...] }
        //   config: { safety_margin_bytes, dram_safety_margin_bytes,
        //          enable_spillover, default_ttl, ollama_poll_s,
        //          grace_period_s, model_cache_ttl_s }
        "Snapshot of GPU/system memory totals, leases, model cache, and config.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.leases",
        |payload: Value| async move { handle_leases(payload).await },
        // Reply: {leases: [Lease]} — each lease carries `bytes` (VRAM) and
        // `dram_bytes` (DRAM portion; 0 for pure-VRAM leases).
        "List all live leases.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.cache",
        |payload: Value| async move { handle_cache(payload).await },
        "List the soft-LRU (service, model) keep-warm cache.",
        "wylde_vram_broker::service",
    );
    register_action_with_meta(
        "vram.evict",
        |payload: Value| async move { handle_evict(payload).await },
        "Force-drop a lease by id; does NOT signal the owning service.",
        "wylde_vram_broker::service",
    );

    // Phase 12.2 — host hardware inventory for the first-run LLM
    // bootstrap. Sampled live on each call (refreshes NVML + sysinfo);
    // the caller gets the same CPU/RAM/disk/GPU/NPU/OS shape regardless
    // of whether the wizard has run yet.
    register_action_with_meta(
        "system.inventory",
        |payload: Value| async move { handle_system_inventory(payload).await },
        // Reply shape: { cpu, memory_total_bytes, memory_available_bytes,
        // disks, gpus, npu, os } — see `inventory::Inventory`.
        "Snapshot of host hardware: CPU brand + cores, RAM, mounted disks, \
         NVIDIA GPUs (via NVML), NPU presence, and OS. Used by the \
         first-run bootstrap LLM to pick models that fit the box.",
        "wylde_vram_broker::service",
    );

    workers::start_background();
}

/// Signal background tasks to stop. Safe to call from a signal handler.
pub fn stop() {
    workers::signal_stop();
}

/// Test-only: reset every singleton in place. Mirrors Python's
/// `_reset_for_tests` — used between cases so each test sees a clean broker
/// without process restarts.
pub fn reset_for_tests() {
    stop();
    for action in [
        "vram.reserve",
        "vram.release",
        "vram.heartbeat",
        "vram.state",
        "vram.leases",
        "vram.cache",
        "vram.evict",
        "system.inventory",
    ] {
        unregister_action(action);
    }
    registry().reset();
    model_cache().reset();
    workers::reset();
    INSTALLED.store(false, Ordering::SeqCst);
}

// ── Action handlers ───────────────────────────────────────────────────

async fn handle_reserve(payload: Value) -> Reply {
    let req: GrantRequest = match serde_json::from_value(payload.clone()) {
        Ok(r) => r,
        Err(e) => {
            return Reply::err(IpcError::new(
                "invalid_request",
                format!("reserve payload decode failed: {e}"),
            ));
        }
    };
    let result = policy::try_grant(req).await;
    match result {
        GrantResult::Ok(v) => {
            // The Python /vram/reserve route returned the bare lease dict on
            // success. Preserve that wire shape; surface `preempted` /
            // `soft_eviction` / `dedup` as sibling fields when present.
            let lease = v.get("lease").cloned().unwrap_or(Value::Null);
            let mut out = lease;
            if let Some(p) = v.get("preempted") {
                if let Value::Object(ref mut m) = out {
                    m.insert("preempted".into(), p.clone());
                }
            }
            if let Some(p) = v.get("soft_eviction") {
                if let Value::Object(ref mut m) = out {
                    m.insert("soft_eviction".into(), p.clone());
                }
            }
            if let Some(p) = v.get("dedup") {
                if let Value::Object(ref mut m) = out {
                    m.insert("dedup".into(), p.clone());
                }
            }
            Reply::ok(out)
        }
        GrantResult::Err(v) => {
            let err = v.get("error").cloned().unwrap_or(Value::Null);
            let code = err
                .get("code")
                .and_then(|c| c.as_str())
                .unwrap_or("unknown")
                .to_owned();
            let message = err
                .get("message")
                .and_then(|c| c.as_str())
                .unwrap_or("reserve failed")
                .to_owned();
            let details = err.get("details").cloned();
            let mut e = IpcError::new(code, message);
            e.details = details;
            Reply::err(e)
        }
    }
}

async fn handle_release(payload: Value) -> Reply {
    let lease_id = payload
        .get("lease_id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_owned();
    match registry().remove(&lease_id) {
        None => Reply::ok(json!({"ok": true, "known": false})),
        Some(lease) => {
            let short = lease.lease_id.get(..8).unwrap_or(&lease.lease_id);
            tracing::info!(
                "vram_broker: release lease={} service={} model={}",
                short,
                lease.service,
                lease.model
            );
            Reply::ok(json!({
                "ok": true,
                "known": true,
                "freed_bytes": lease.bytes,
            }))
        }
    }
}

async fn handle_heartbeat(payload: Value) -> Reply {
    let lease_id = payload
        .get("lease_id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_owned();
    let ttl = payload
        .get("ttl")
        .and_then(|v| v.as_f64())
        .unwrap_or(Config::get().default_ttl);
    match registry().touch(&lease_id, ttl) {
        None => Reply::err(IpcError::new(
            "not_found",
            "lease not found or already reaped",
        )),
        Some(lease) => Reply::ok(json!({
            "lease_id": lease.lease_id,
            "expires_at": lease.expires_at,
        })),
    }
}

async fn handle_state(_payload: Value) -> Reply {
    refresh_nvml();
    Reply::ok(workers::state_snapshot())
}

async fn handle_leases(_payload: Value) -> Reply {
    let leases: Vec<Value> = registry()
        .all_leases()
        .iter()
        .map(|l| serde_json::to_value(l).unwrap_or(Value::Null))
        .collect();
    Reply::ok(json!({ "leases": leases }))
}

async fn handle_cache(_payload: Value) -> Reply {
    let cfg = Config::get();
    let mut entries = model_cache().all();
    entries.sort_by(|a, b| {
        b.last_used
            .partial_cmp(&a.last_used)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    let now = now_secs();
    let entries_arr: Vec<Value> = entries
        .into_iter()
        .map(|e| {
            let warm_for = (cfg.model_cache_ttl_s - (now - e.last_used)).max(0.0);
            json!({
                "service": e.service,
                "model": e.model,
                "bytes": e.bytes,
                "last_used": e.last_used,
                "warm_for": warm_for,
            })
        })
        .collect();
    Reply::ok(json!({
        "ttl_s": cfg.model_cache_ttl_s,
        "entries": entries_arr,
    }))
}

async fn handle_evict(payload: Value) -> Reply {
    let lease_id = payload
        .get("lease_id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_owned();
    match registry().remove(&lease_id) {
        None => Reply::err(IpcError::new("not_found", "lease not found")),
        Some(lease) => Reply::ok(json!({
            "ok": true,
            "freed_bytes": lease.bytes,
        })),
    }
}

async fn handle_system_inventory(_payload: Value) -> Reply {
    Reply::ok(inventory::handle_inventory())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lock::guard;

    const GB: u64 = 1024 * 1024 * 1024;

    fn fresh() {
        reset_for_tests();
        registry().set_gpu(16 * GB, 0, "TestGPU");
        install(false);
    }

    #[tokio::test(flavor = "current_thread")]
    async fn reserve_returns_lease_shape() {
        let _g = guard().await;
        fresh();
        let reply = handle_reserve(json!({
            "service": "wylde-caption",
            "model": "florence-2",
            "bytes": 4 * GB,
            "priority": 40,
            "ttl": 60.0,
        }))
        .await;
        assert!(reply.ok);
        assert_eq!(reply.data["service"], "wylde-caption");
        assert_eq!(reply.data["bytes"], 4 * GB);
        assert!(reply.data["lease_id"].is_string());
    }

    #[tokio::test(flavor = "current_thread")]
    async fn release_known_and_unknown() {
        let _g = guard().await;
        fresh();
        let granted = handle_reserve(json!({
            "service": "wylde-rag",  // wylde-check: dead-ref-ok
            "model": "reranker",
            "bytes": 2 * GB,
            "priority": 60,
        }))
        .await;
        let lid = granted.data["lease_id"].as_str().unwrap().to_owned();

        let r1 = handle_release(json!({"lease_id": lid.clone()})).await;
        assert!(r1.ok);
        assert_eq!(r1.data["known"], true);
        assert_eq!(r1.data["freed_bytes"], 2 * GB);

        let r2 = handle_release(json!({"lease_id": lid})).await;
        assert_eq!(r2.data["known"], false);
    }

    #[tokio::test(flavor = "current_thread")]
    async fn heartbeat_extends_and_404() {
        let _g = guard().await;
        fresh();
        let granted = handle_reserve(json!({
            "service": "wylde-caption",
            "model": "m",
            "bytes": GB,
            "priority": 40,
            "ttl": 5.0,
        }))
        .await;
        let lid = granted.data["lease_id"].as_str().unwrap().to_owned();
        let first_exp = granted.data["expires_at"].as_f64().unwrap();

        std::thread::sleep(std::time::Duration::from_millis(20));
        let hb = handle_heartbeat(json!({"lease_id": lid, "ttl": 60.0})).await;
        assert!(hb.ok);
        assert!(hb.data["expires_at"].as_f64().unwrap() > first_exp);

        let miss = handle_heartbeat(json!({"lease_id": "nope"})).await;
        assert!(!miss.ok);
        assert_eq!(miss.error.unwrap().code, "not_found");
    }

    #[tokio::test(flavor = "current_thread")]
    async fn state_includes_gpu_and_leases() {
        let _g = guard().await;
        fresh();
        let _ = handle_reserve(json!({
            "service": "wylde-caption",
            "model": "x",
            "bytes": 2 * GB,
            "priority": 40,
        }))
        .await;
        let s = handle_state(Value::Null).await;
        assert!(s.ok);
        assert_eq!(s.data["gpu"]["total_bytes"], 16 * GB);
        assert_eq!(s.data["gpu"]["reserved_bytes"], 2 * GB);
        let leases = s.data["leases"].as_array().unwrap();
        assert!(leases.iter().any(|l| l["service"] == "wylde-caption"));
        let priorities: Vec<i64> = s.data["by_service"]
            .as_array()
            .unwrap()
            .iter()
            .map(|r| r["priority"].as_i64().unwrap())
            .collect();
        let mut sorted = priorities.clone();
        sorted.sort_by(|a, b| b.cmp(a));
        assert_eq!(priorities, sorted);
    }

    #[tokio::test(flavor = "current_thread")]
    async fn evict_unknown_is_not_found() {
        let _g = guard().await;
        fresh();
        let r = handle_evict(json!({"lease_id": "nope"})).await;
        assert!(!r.ok);
        assert_eq!(r.error.unwrap().code, "not_found");
    }
}
